src/coreclr/tools/Common/TypeSystem/Interop/IL/MarshalHelpers.cs

```diff
@@ -521,7 +521,7 @@ public static partial class MarshalHelpers
                 else
                     return MarshallerKind.Invalid;
             }
-            else if (type.IsDelegate)
+            else if (type.IsDelegate || InteropTypes.IsSystemDelegate(context, type) || InteropTypes.IsSystemMulticastDelegate(context, type))
```

Comment:
`System.Delegate` and `System.MulticastDelegate` marshaling is not AOT friendly. We cannot tell the delegate type to use during AOT compilation for these cases. The marshaling logic is suppressing the AOT warning for the delegates, but this suppression is not correct for the abstract delegate types. 

This change also needs a matching change in the marshalling logic to avoid suppressing the AOT compatibility warning for these cases.

Reply:
Short version: I miss some part of puzzle about who produce AOT warnings. Please advice.

## Longer version

I look at places where mentioned `MulticastDelegate` in `src/coreclr/tools/`
https://github.com/dotnet/runtime/blob/b2dc37ba181a7fa4427e717eab819ba3543d0ae4/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs#L2371
and others

Then for `IsDelegate` in same folder. Since this checks for something derived from MulticastDelegate I clearly do not see 
https://github.com/dotnet/runtime/blob/b2dc37ba181a7fa4427e717eab819ba3543d0ae4/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedInteropStubManager.cs#L109
and others

`DelegateMarshaller` relies on `PInvokeMarshal` which does not annotated for AOT warnings.
https://github.com/dotnet/runtime/blob/6e286dd75eccf607d2b17795bc23b6a82bf0d97d/src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs#L2023-L2027

`PInvokeMarshal` class itself appears to not care at all about marshalling. I assume this is low-level logic.
https://github.com/dotnet/runtime/blob/b2dc37ba181a7fa4427e717eab819ba3543d0ae4/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/PInvokeMarshal.cs#L46

Reply:
AOT warnings are produced by calling methods annotated with `RequiresUnreferencedCode`. I think you may want to create a new methods on `PInvokeMarshal` that are annotated with `RequiresUnreferencedCode` and just forward to the unannotated methods, and call these methods for abstract delegates.

Reply:
This will produce a warning, but it won't be a particularly useful warning (it will produce a warning saying that a compiler-generated method is calling a method annotated `RequiresUnreferencedCode`, which is not particularly actionable).

The proper fix would be to add handling of this somewhere around here (and a new warning code):

https://github.com/dotnet/runtime/blob/d0631be3a1b9e5a8adf16f4885778179a8ed365b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs#L2178-L2203

There's a pull request in flight that moves the warning codes to someplace else, so I would wait for that (#63230) to land first.

Reply:
I assume that plan was to revert last commit which generate annotated code and made changes in analysis. Force push changes.